src/persist-client/src/paths.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{ShardId, WriterId};
+
+/// An opaque identifier for an individual blob of a persist durable TVC (aka shard).
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct PartId(pub(crate) [u8; 16]);
+
+impl std::fmt::Display for PartId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "p{}", Uuid::from_bytes(self.0))
+    }
+}
+
+impl std::fmt::Debug for PartId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "PartId({})", Uuid::from_bytes(self.0))
+    }
+}
+
+impl FromStr for PartId {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        parse_id('p', "PartId", s).map(PartId)
+    }
+}
+
+impl PartId {
+    pub(crate) fn new() -> Self {
+        PartId(*Uuid::new_v4().as_bytes())
+    }
+}
+
+/// Partially encoded path used in [mz_persist::location::Blob] storage.
+/// Composed of a [crate::write::WriterId] and [crate::paths::PartId]. Can
+/// be completed with a [crate::ShardId] to form a full [crate::paths::BlobKey].
+///
+/// Used to reduce the bytes needed to refer to a blob key in memory and
+/// in persistent state, All access to blobs are always within the context
+/// of an individual shard.
+#[derive(Clone, Debug, PartialEq)]
+pub struct PartialBlobKey(pub(crate) String);
+
+impl PartialBlobKey {
+    pub fn new(writer_id: &WriterId) -> PartialBlobKey {
+        PartialBlobKey(format!("{}/{}", writer_id, PartId::new()))
+    }
+
+    pub fn complete(&self, shard_id: &ShardId) -> BlobKey {
+        BlobKey(format!("{}/{}", shard_id, self))
+    }
+}
+
+impl std::fmt::Display for PartialBlobKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+/// Fully encoded path used in [mz_persist::location::Blob] storage.
+/// Composed of a [crate::ShardId], [crate::write::WriterId] and
+/// [crate::paths::PartId].
+///
+/// Use when directly interacting with a [mz_persist::location::Blob],
+/// otherwise use [crate::paths::PartialBlobKey] to refer to a blob without
+/// needing to copy the [crate::ShardId].
+#[derive(Clone, Debug, PartialEq)]
+pub struct BlobKey(pub(crate) String);
+
+impl std::fmt::Display for BlobKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&self.0)
+    }
+}
+
+impl BlobKey {
+    pub fn parse_ids(key: &str) -> Result<(Option<ShardId>, Option<WriterId>, PartId), String> {
+        let ids = key.split('/').collect::<Vec<_>>();
+
+        if ids.is_empty() {
+            return Err("empty blob key".to_string());
+        } else if ids.len() > 3 {
+            return Err(format!(
+                "invalid blob key, expected <= 3 ids, found {}: {:?}",
+                ids.len(),
+                ids
+            ));
+        }
+
+        let mut shard_id = None;
+        let mut writer_id = None;
+        let mut part_id = None;
+        for id in ids {
```

Comment:
I have a mild preference to enforce that this is either len 1 or len 3 and to change the return type to `Result<(Option<(ShardId, WriterId)>, PartId), String>` with a TODO that we can delete the len 1 support (and remove the Option wrapper for ShardId and WriteId once we've had to wipe cloud's environments